# ETL Proyecto Individual ML Ops STEAM

In [1]:
# Importo las librerias que considero necesarias para hacer la extración de los Data Set
import json
import ast
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Ruta de ubicación para los datasets:
# 'E:\\AAADATOS\Henry\\AA_Data_Science\\MATERIAL_PI\\DATA_PI MLOps - STEAM\\Data_sets\\australian_user_reviews.json'
# 'E:\\AAADATOS\Henry\\AA_Data_Science\\MATERIAL_PI\\DATA_PI MLOps - STEAM\\Data_sets\\australian_users_items.json'
# 'E:\\AAADATOS\Henry\\AA_Data_Science\\MATERIAL_PI\\DATA_PI MLOps - STEAM\\Data_sets\\output_steam_games.json'

## Data Set australian_user_reviews.json

In [3]:
# Se inicializa una lista vacia donde alamcenare los datos del archivo .JSON
filas = []
# Se abre el archivo .JSON con el modulo 'with'
with open('E:\\AAADATOS\\Henry\\AA_Data_Science\\MATERIAL_PI\\DATA_PI MLOps - STEAM\\Data_sets\\australian_user_reviews.json', 'r', encoding='utf-8') as archivo:
    for line in archivo.readlines():
        filas.append(ast.literal_eval(line))

In [4]:
# Se inicializa el primer DataFrame a partir de la lista donde se extrajo el archivo .JSON
df_reviews = pd.DataFrame(filas)
df_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

Se evidencia que la columna "reviews" contiene listas como dato así que se tiene que proceder a expandirla

In [54]:
df_reviews = df_reviews.explode('reviews').reset_index() # Se redefine el DatFrame con una descomposición del campo para desanidarla y se reinicia el conteo del índice
df_reviews = df_reviews.drop(columns='index') # Se elimina el campo 'index'
df_reviews = pd.concat([df_reviews, pd.json_normalize(df_reviews['reviews'])], axis=1) # Se normalizan los datos JSON de la columna 'reviews' aplanando los datos, convirtiendolos en un formato tabular y concatenando de manera horizontal el DatFrame original con el del contendio de 'reviews' 
df_reviews.head()

KeyError: 'reviews'

### Se empieza a analizar y transformar los datos del DataFrame

In [ ]:
# Información general del DataFrame
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59333 entries, 0 to 59332
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      59333 non-null  object
 1   user_url     59333 non-null  object
 2   reviews      59305 non-null  object
 3   funny        59305 non-null  object
 4   posted       59305 non-null  object
 5   last_edited  59305 non-null  object
 6   item_id      59305 non-null  object
 7   helpful      59305 non-null  object
 8   recommend    59305 non-null  object
 9   review       59305 non-null  object
dtypes: object(10)
memory usage: 4.5+ MB


NULOS

In [ ]:
# Conteo de Nulos
nulos = df_reviews.isna().sum()
nulos

user_id         0
user_url        0
reviews        28
funny          28
posted         28
last_edited    28
item_id        28
helpful        28
recommend      28
review         28
dtype: int64

In [ ]:
# Visualizar los valores Nulos
df_nulos = df_reviews[df_reviews.isnull().any(axis=1)] # Se crea un nuevo DatFrame donde se filtra si hay algun valor nulo en la fila a partir del DataFrame orginal
df_nulos

user_id                                           user_url  \
137                gdxsd                 http://steamcommunity.com/id/gdxsd   
177    76561198094224872  http://steamcommunity.com/profiles/76561198094...   
2559   76561198021575394  http://steamcommunity.com/profiles/76561198021...   
10080            cmuir37               http://steamcommunity.com/id/cmuir37   
13767          Jaysteeny             http://steamcommunity.com/id/Jaysteeny   
15493             ML8989                http://steamcommunity.com/id/ML8989   
19184  76561198079215291  http://steamcommunity.com/profiles/76561198079...   
20223  76561198079342142  http://steamcommunity.com/profiles/76561198079...   
25056  76561198061996985  http://steamcommunity.com/profiles/76561198061...   
26257  76561198108286351  http://steamcommunity.com/profiles/76561198108...   
26674     CallTripleZero        http://steamcommunity.com/id/CallTripleZero   
28444       Priceless612          http://steamcommunity.com/id/Priceless612   
28452      diabolical666         http://steamcommunity.com/id/diabolical666   
28975         DrScottnik            http://steamcommunity.com/id/DrScottnik   
31955  76561198110176420  http://steamcommunity.com/profiles/76561198110...   
32819         jojoPL1987            http://steamcommunity.com/id/jojoPL1987   
33785        Kylee_Kylie           http://steamcommunity.com/id/Kylee_Kylie   
39279             _samie                http://steamcommunity.com/id/_samie   
39357   FifthExclamation      http://steamcommunity.com/id/FifthExclamation   
39420             sbenji                http://steamcommunity.com/id/sbenji   
40661     SpatialWheel14        http://steamcommunity.com/id/SpatialWheel14   
41777  76561198065238937  http://steamcommunity.com/profiles/76561198065...   
44384  76561198071075080  http://steamcommunity.com/profiles/76561198071...   
50718  76561198063489150  http://steamcommunity.com/profiles/76561198063...   
53323   thecommunistbear      http://steamcommunity.com/id/thecommunistbear   
55425  76561198085883839  http://steamcommunity.com/profiles/76561198085...   
57578  76561198100396147  http://steamcommunity.com/profiles/76561198100...   
57766  76561198102423654  http://steamcommunity.com/profiles/76561198102...   

      reviews funny posted last_edited item_id helpful recommend review  
137       NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
177       NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
2559      NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
10080     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
13767     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
15493     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
19184     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
20223     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
25056     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
26257     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
26674     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
28444     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
28452     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
28975     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
31955     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
32819     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
33785     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
39279     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
39357     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
39420     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
40661     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
41777     NaN   NaN    NaN         NaN     NaN     NaN       NaN    NaN  
44384     NaN   NaN    

En base al DataFrame anterior se puede afirmar que todos los Nulos corresponden estas filas donde no esta la información relevante para mis objetivos con las consignas que se me asignaron por lo cual se procedera a ser eliminados

In [ ]:
# Elimino las filas con valores nulos del DataFrame con el método 'dropna()'
df_reviews.dropna(inplace=True)
# Conteo de Nulos nuevamente para verificar
nuevos_nulos = df_reviews.isna().sum()
nuevos_nulos

user_id        0
user_url       0
reviews        0
funny          0
posted         0
last_edited    0
item_id        0
helpful        0
recommend      0
review         0
dtype: int64

De esta manera se han eliminado los valores Nulos del DataFrame hasta este punto de la transformación

#### Se procede a evaluar el contenido de los campos y determinar su relevancia dentro de los obejtivos

In [ ]:
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 59305 entries, 0 to 59332
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      59305 non-null  object
 1   user_url     59305 non-null  object
 2   reviews      59305 non-null  object
 3   funny        59305 non-null  object
 4   posted       59305 non-null  object
 5   last_edited  59305 non-null  object
 6   item_id      59305 non-null  object
 7   helpful      59305 non-null  object
 8   recommend    59305 non-null  object
 9   review       59305 non-null  object
dtypes: object(10)
memory usage: 5.0+ MB


In [ ]:
# Se crea una variable que contiene los valores unicos del campo 'funny'
valores_funny = df_reviews['funny'].unique() 
valores_funny

array(['', '3 people found this review funny',
       '1 person found this review funny',
       '2 people found this review funny',
       '5 people found this review funny',
       '17 people found this review funny',
       '524 people found this review funny',
       '4 people found this review funny',
       '12 people found this review funny',
       '6 people found this review funny',
       '7 people found this review funny',
       '103 people found this review funny',
       '16 people found this review funny',
       '34 people found this review funny',
       '9 people found this review funny',
       '45 people found this review funny',
       '11 people found this review funny',
       '97 people found this review funny',
       '207 people found this review funny',
       '718 people found this review funny',
       '23 people found this review funny',
       '100 people found this review funny',
       '8 people found this review funny',
       '19 people found this rev

In [ ]:
# Se crea otra variable para contar cuantos valores hay que no sean strings vacios ("")
cant_funny = df_reviews[df_reviews['funny'] != ''].count()['funny']
cant_funny

8151

In [ ]:
# Se hace una operación matematica para determinar que porcentaje representa los valores de funny completos dentro del conjunto del DataFrame
total_filas_df = 59305
print(f'La cantidad de valores que no son strings vacios en el campo "funny" representan un {round((cant_funny/total_filas_df)*100, 2)}%')

La cantidad de valores que no son strings vacios en el campo "funny" representan un 13.74%


Después del proceso anterior se puede determinar que los valores que se podrían considerar útiles del campo "funny" apenas representan un 13.74% dentro de todo el DataFrame por lo cual este campo en su conjunto se considerará presindible

In [ ]:
# Se crea una variable que contiene los valores unicos del campo 'last_edited'
valores_last_edited = df_reviews['last_edited'].unique() 
valores_last_edited

array(['', 'Last edited December 5, 2015.',
       'Last edited November 25, 2013.', ...,
       'Last edited April 17, 2014.', 'Last edited May 10, 2014.',
       'Last edited August 15, 2014.'], dtype=object)

In [ ]:
# Se crea otra variable para contar cuantos valores hay que no sean strings vacios ("")
cant_last_edited = df_reviews[df_reviews['last_edited'] != ''].count()['last_edited']
cant_last_edited

6140

In [ ]:
# Se hace una operación matematica para determinar que porcentaje representa los valores de "last_edited" completos dentro del conjunto del DataFrame
print(f'La cantidad de valores que no son strings vacios en el campo "last_edited" representan un {round((cant_last_edited/total_filas_df)*100, 2)}%')

La cantidad de valores que no son strings vacios en el campo "last_edited" representan un 10.35%


Después del proceso anterior se puede determinar que los valores que se podrían considerar útiles del campo "last_edited" apenas representan un 10.35% dentro de todo el DataFrame por lo cual este campo en su conjunto se considerará presindible

In [ ]:
# Se crea una variable que contiene los valores unicos del campo 'helpful'
valores_helpful = df_reviews['helpful'].unique() 
valores_helpful

array(['No ratings yet',
       '15 of 20 people (75%) found this review helpful',
       '0 of 1 people (0%) found this review helpful', ...,
       '21 of 34 people (62%) found this review helpful',
       '64 of 73 people (88%) found this review helpful',
       '24 of 34 people (71%) found this review helpful'], dtype=object)

In [ ]:
# Se crea otra variable para contar cuantos valores hay que no sean "No ratings yet"
cant_helpful = df_reviews[df_reviews['helpful'] != 'No ratings yet'].count()['helpful']
cant_helpful

29137

In [ ]:
# Se hace una operación matematica para determinar que porcentaje representa los valores de "helpful" completos dentro del conjunto del DataFrame
print(f'La cantidad de valores que no son "No ratings yet" vacios en el campo "helpful" representan un {round((cant_helpful/total_filas_df)*100, 2)}%')

La cantidad de valores que no son "No ratings yet" vacios en el campo "helpful" representan un 49.13%


Ahora que se posee mas información sobre el campo "helpful" se puede determinar que es presindible no solo por que su información relevante represente un poco menos del 50%, sino también por que esta información 'útil' seria dificil de cuantificar y llevar a escala además en los requerimientos se indica que el campo mas relevante para el objetivo es "review" por lo cual este campo tambien se presindé.

In [ ]:
# Se crea una variable que contiene los valores unicos del campo 'recommend'
valores_recommend = df_reviews['recommend'].unique() 
valores_recommend

array([True, False], dtype=object)

Con el proceso anterior se cerciora que el campo "recommend" contiene información relevante en todas las filas del DataFrame

#### Fechas

In [ ]:
# Se elimina la cadena 'Posted' de todos las filas del campo "posted"
df_reviews['posted'] = df_reviews['posted'].str.replace('Posted', '')
df_reviews.head(2)

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

                                             reviews funny  \
0  {'funny': '', 'posted': 'Posted November 5, 20...         
1  {'funny': '', 'posted': 'Posted July 15, 2011....         

               posted last_edited item_id         helpful recommend  \
0   November 5, 2011.                1250  No ratings yet      True   
1      July 15, 2011.               22200  No ratings yet      True   

                                              review  
0  Simple yet with great replayability. In my opi...  
1               It's unique and worth a playthrough.

In [ ]:
# Se crea un nuevo campo llamado "Date" donde se almacenara el contenido de "posted" convertido a formato datetime
df_reviews['Date'] = pd.to_datetime(df_reviews['posted'], dayfirst=True, errors='coerce') # Con estos atributos se logra primero que se reconozca adecuadamente el formato de la fecha en el campo "posted" y segundo que lo que no se reconozca como datetime se convierta en una NaT "Not a Time"
df_reviews.head(2)

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

                                             reviews funny  \
0  {'funny': '', 'posted': 'Posted November 5, 20...         
1  {'funny': '', 'posted': 'Posted July 15, 2011....         

               posted last_edited item_id         helpful recommend  \
0   November 5, 2011.                1250  No ratings yet      True   
1      July 15, 2011.               22200  No ratings yet      True   

                                              review       Date  
0  Simple yet with great replayability. In my opi... 2011-11-05  
1               It's unique and worth a playthrough. 2011-07-15

In [ ]:
# Se crea una variable para contabilizar la cantidad NaT's que hay en el nuevo campo "Date" del DataFrame
cant_Not_Date = df_reviews['Date'].isna().sum()
cant_Not_Date

10119

In [ ]:
# Se crea un nuevo campor llamado "Year" donde se almacenara solamente el año de la review
df_reviews['Year'] = pd.to_datetime(df_reviews['Date']).dt.year
df_reviews.head(2)

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

                                             reviews funny  \
0  {'funny': '', 'posted': 'Posted November 5, 20...         
1  {'funny': '', 'posted': 'Posted July 15, 2011....         

               posted last_edited item_id         helpful recommend  \
0   November 5, 2011.                1250  No ratings yet      True   
1      July 15, 2011.               22200  No ratings yet      True   

                                              review       Date    Year  
0  Simple yet with great replayability. In my opi... 2011-11-05  2011.0  
1               It's unique and worth a playthrough. 2011-07-15  2011.0

Debido a que se cuenta con valores NaT en el campo "Date" el output del campo "Year" esta en formato float por lo cual se van a tranformar a formato "int"

In [ ]:
# Se cuenta la cantidad de frecuencia de valores unicos que posee el campo "Year"
fecuencia_valores_Year = df_reviews['Year'].value_counts()
fecuencia_valores_Year

Year
2014.0    22111
2015.0    18470
2013.0     6792
2012.0     1215
2011.0      532
2010.0       66
Name: count, dtype: int64

In [ ]:
# Se crea una variable para contabilizar la cantidad NaT's que hay en el nuevo campo "Year" del DataFrame
cant_Not_Year = df_reviews['Year'].isna().sum()
cant_Not_Year

10119

Con las tranformaciones anteriores se puede establecer que la cantidad de filas donde faltan unicamente el año en el campo "posted" son 10119 y es por esta razón que hay nulos al momento de transformalos a formatos de fecha

In [ ]:
# Se reemplaza los valores NaT del campo "Year" por el año 2016
df_reviews['Year'].fillna(2016, inplace=True)
cant_Not_Year = df_reviews['Year'].isna().sum() # Se vuelve a contar los valores Nulos del campo
cant_Not_Year

0

In [ ]:
# Se cuenta la cantidad de frecuencia de valores unicos que posee el campo "Year" nuevamente
fecuencia_valores_Year = df_reviews['Year'].value_counts()
fecuencia_valores_Year

Year
2014.0    22111
2015.0    18470
2016.0    10119
2013.0     6792
2012.0     1215
2011.0      532
2010.0       66
Name: count, dtype: int64

Esta decisión se justifica en un análisis externo de sobre Steam en la cual se encuentra que las reviews que no llevan el año en la fecha de posteo es por que fueron realizadas en el año en curso de esta manera como el último año que se tenía registro en el campo "posted" era el "2015" se asume que el año en curso es el "2016" y este pasa a reemplazar los valores Nulos que se tenian en estos campos de fechas

In [ ]:
# Se transforma los datos del campo "Year" de tipo float a tipo int
df_reviews['Year'] = df_reviews['Year'].astype(int)
df_reviews.head(2)

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

                                             reviews funny  \
0  {'funny': '', 'posted': 'Posted November 5, 20...         
1  {'funny': '', 'posted': 'Posted July 15, 2011....         

               posted last_edited item_id         helpful recommend  \
0   November 5, 2011.                1250  No ratings yet      True   
1      July 15, 2011.               22200  No ratings yet      True   

                                              review       Date  Year  
0  Simple yet with great replayability. In my opi... 2011-11-05  2011  
1               It's unique and worth a playthrough. 2011-07-15  2011

In [ ]:
# Se elimina el punto '.' de todos las filas del campo "posted"
df_reviews['posted'] = df_reviews['posted'].str.replace('.', '')
df_reviews.head(2)

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

                                             reviews funny             posted  \
0  {'funny': '', 'posted': 'Posted November 5, 20...         November 5, 2011   
1  {'funny': '', 'posted': 'Posted July 15, 2011....            July 15, 2011   

  last_edited item_id         helpful recommend  \
0                1250  No ratings yet      True   
1               22200  No ratings yet      True   

                                              review       Date  Year  
0  Simple yet with great replayability. In my opi... 2011-11-05  2011  
1               It's unique and worth a playthrough. 2011-07-15  2011

Se procede a modificar el campo "posted" para poder añadir el año "2016" donde no lo tiene y de esta manera poder completar todos los valores del campo "Date"

In [ ]:
patron_4_digitos = r'(\d{4})$' # Se crea una variable para identificar los valores donde no haya un patrón de cuatro dígitos consecutivos
df_reviews.loc[~df_reviews['posted'].str.contains(patron_4_digitos, regex=True), 'posted'] += ', 2016' # Se crea una máscara booleana que es True para las filas donde no hay un patrón de 4 números consecutivos al final

C:\Users\DavidssonPC\AppData\Local\Temp\ipykernel_7448\182169166.py:2: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  df_reviews.loc[~df_reviews['posted'].str.contains(patron_4_digitos, regex=True), 'posted'] += ', 2016' # Se crea una máscara booleana que es True para las filas donde no hay un patrón de 4 números consecutivos al final


El siguiente paso es hacer nuevamente el campo "Date" pero esta vez  se completaran todos los campos así no quedando valores Nulos

In [ ]:
# Se crea elimina el anterior campo "Date" y se transforma nuevamente
df_reviews = df_reviews.drop(columns='Date')
df_reviews['Date'] = pd.to_datetime(df_reviews['posted'], dayfirst=True, errors='coerce') # Con estos atributos se logra primero que se reconozca adecuadamente el formato de la fecha en el campo "posted" y segundo que lo que no se reconozca como datetime se convierta en una NaT "Not a Time"

In [ ]:
# Se nuevamente la variable para contabilizar la cantidad NaT's que hay en el nuevo campo "Date" del DataFrame
cant_Not_Date = df_reviews['Date'].isna().sum()
cant_Not_Date

0

Esta vez ya no hay valores Nulos en los campos de fechas

#### Eliminar campos no relevantes

In [ ]:
# Se crea una lista con el nombre de los campos que se van a eliminar del DataFrame
campos_out = ['user_url','reviews','funny','posted','last_edited','helpful']

In [ ]:
df_reviews = df_reviews.drop(columns=campos_out)
df_reviews.head(2)

user_id item_id recommend  \
0  76561197970982479    1250      True   
1  76561197970982479   22200      True   

                                              review  Year       Date  
0  Simple yet with great replayability. In my opi...  2011 2011-11-05  
1               It's unique and worth a playthrough.  2011 2011-07-15

En este punto que ya se tienen los campos relevantes se procede a normalizar los nombre de estos

In [ ]:
df_reviews = df_reviews.rename(columns={'user_id': 'User_Id'})
df_reviews = df_reviews.rename(columns={'item_id': 'Item_Id'})
df_reviews = df_reviews.rename(columns={'recommend': 'Recommend'})
df_reviews = df_reviews.rename(columns={'review': 'Review'})
df_reviews.head(2)

User_Id Item_Id Recommend  \
0  76561197970982479    1250      True   
1  76561197970982479   22200      True   

                                              Review  Year       Date  
0  Simple yet with great replayability. In my opi...  2011 2011-11-05  
1               It's unique and worth a playthrough.  2011 2011-07-15

#### Se crea la columna 'sentiment_analysis' aplicando análisis de sentimiento con NLP

Se importa la libreria nltk y se descarga el lexicon 'vader_lexicon' necesario para realizar el análisis de sentimientos

In [ ]:
import nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\DavidssonPC\AppData\Roaming\nltk_data...


Se crea la función que analizara los valores de la columna "Review" para darles un valor númerico según el análisis de sentimientos

In [ ]:
# Se inicializa el analizador de sentimientos 
AS = SentimentIntensityAnalyzer()

# '0' si es malo 
# '1' si es neutral 
# '2' si es positivo

# Funcion para asignar valores de acuerdo a la escala
def puntaje_sentimiento(texto): # La variable de entrada de mi función seran los valores en formato str de la columna "Review"
    if pd.isnull(texto) or texto == '': # Evalua si el valor es un Nulo o un string vacío
        return 1 # Si es así retorna 1 que es un valor neutral
    elif isinstance(texto, str): # Verifica que el "texto" se una instancia del tipo str
        sentimiento = AS.polarity_scores(texto) # Se utiliza el analizador de sentimientos 'AS' para obtener un diccionario de puntajes de polarida y la función polarity_scores devuelve un diccionario que contiene puntajes para la polaridad positiva, negativa, neutra y un puntaje compuesto que resume la polaridad general del texto
        puntaje_compuesto = sentimiento['compound'] # Se extrae el puntaje compuesto del diccionario de sentimientos
        if puntaje_compuesto >= -0.05:
            return 2 # es positivo
        elif puntaje_compuesto <= -0.05:
            return 0 # es malo
        else:
            return 1 # es neutral
    else:
        return 1 # Si no cumple ninguna condición es neutral

In [ ]:
# Se convierte la columna 'Review' a formato str
df_reviews['Review'] = df_reviews['Review'].astype(str)

# Se aplica la función puntaje_sentimiento a la columna 'Review' creando una nueva columna llamada 'Sentiment_Analysis' que contendra los puntajes de este análisis
df_reviews['Sentiment_Analysis'] = df_reviews['Review'].apply(puntaje_sentimiento)
df_reviews.head(2)

# Tardo 20.7 seg

User_Id Item_Id Recommend  \
0  76561197970982479    1250      True   
1  76561197970982479   22200      True   

                                              Review  Year       Date  \
0  Simple yet with great replayability. In my opi...  2011 2011-11-05   
1               It's unique and worth a playthrough.  2011 2011-07-15   

   Sentiment_Analysis  
0                   2  
1                   2

Para este punto ya no es necesaria la columna "Review" y tal como se pide en las consignas esta ya cumplió su función y fue reemplazada por "Sentiment_Analysis" por lo caul va a ser eliminada

In [ ]:
df_reviews = df_reviews.drop(columns=['Review']) # Se elimina el campo 'Review'
df_reviews.head(2)

User_Id Item_Id Recommend  Year       Date  Sentiment_Analysis
0  76561197970982479    1250      True  2011 2011-11-05                   2
1  76561197970982479   22200      True  2011 2011-07-15                   2

In [ ]:
df_reviews

User_Id Item_Id Recommend  Year       Date  \
0      76561197970982479    1250      True  2011 2011-11-05   
1      76561197970982479   22200      True  2011 2011-07-15   
2      76561197970982479   43110      True  2011 2011-04-21   
3                js41637  251610      True  2014 2014-06-24   
4                js41637  227300      True  2013 2013-09-08   
...                  ...     ...       ...   ...        ...   
59328  76561198312638244      70      True  2016 2016-07-10   
59329  76561198312638244  362890      True  2016 2016-07-08   
59330        LydiaMorley  273110      True  2016 2016-07-03   
59331        LydiaMorley     730      True  2016 2016-07-20   
59332        LydiaMorley     440      True  2016 2016-07-02   

       Sentiment_Analysis  
0                       2  
1                       2  
2                       2  
3                       2  
4                       2  
...                   ...  
59328                   2  
59329                   2  
59330                   2  
59331                   2  
59332                   2  

[59305 rows x 6 columns]

### Se guarda el DataFrame en formatos mas ligeros

In [ ]:
df_reviews.to_csv('user_reviews.csv', index=False) # Se guarda el DataFrame en formato CSV

In [ ]:
# Se convierte el archivo CSV a parquet bajo la compresión "gzip"
ruta_csv_reviews = 'E:\\AAADATOS\\Henry\\AA_Data_Science\\MATERIAL_PI\\PI_ML_OPS_STEAM_DSFT17\\user_reviews.csv' # Se crea una variable con la ruta del archivo CSV
df_reviews_temp = pd.read_csv(ruta_csv_reviews) # Se lee ese CSV en un nuevo DataFrame temporal para segurar como estan los datos
df_reviews_temp

User_Id  Item_Id  Recommend  Year        Date  \
0      76561197970982479     1250       True  2011  2011-11-05   
1      76561197970982479    22200       True  2011  2011-07-15   
2      76561197970982479    43110       True  2011  2011-04-21   
3                js41637   251610       True  2014  2014-06-24   
4                js41637   227300       True  2013  2013-09-08   
...                  ...      ...        ...   ...         ...   
59300  76561198312638244       70       True  2016  2016-07-10   
59301  76561198312638244   362890       True  2016  2016-07-08   
59302        LydiaMorley   273110       True  2016  2016-07-03   
59303        LydiaMorley      730       True  2016  2016-07-20   
59304        LydiaMorley      440       True  2016  2016-07-02   

       Sentiment_Analysis  
0                       2  
1                       2  
2                       2  
3                       2  
4                       2  
...                   ...  
59300                   2  
59301                   2  
59302                   2  
59303                   2  
59304                   2  

[59305 rows x 6 columns]

Se guarada el DataFrame temporal en formato comprimido gzip

In [ ]:
df_reviews_temp.to_csv('user_reviews.gzip', compression='gzip', index=False)

In [ ]:
# Se verifica que el archivo gzip comprimido pueda ser leido de manera efectiva y que no hayan problemas en los datos que este arroja
df_reviews_parquet = pd.read_csv('user_reviews.gzip', compression='gzip')
df_reviews_parquet

User_Id  Item_Id  Recommend  Year        Date  \
0      76561197970982479     1250       True  2011  2011-11-05   
1      76561197970982479    22200       True  2011  2011-07-15   
2      76561197970982479    43110       True  2011  2011-04-21   
3                js41637   251610       True  2014  2014-06-24   
4                js41637   227300       True  2013  2013-09-08   
...                  ...      ...        ...   ...         ...   
59300  76561198312638244       70       True  2016  2016-07-10   
59301  76561198312638244   362890       True  2016  2016-07-08   
59302        LydiaMorley   273110       True  2016  2016-07-03   
59303        LydiaMorley      730       True  2016  2016-07-20   
59304        LydiaMorley      440       True  2016  2016-07-02   

       Sentiment_Analysis  
0                       2  
1                       2  
2                       2  
3                       2  
4                       2  
...                   ...  
59300                   2  
59301                   2  
59302                   2  
59303                   2  
59304                   2  

[59305 rows x 6 columns]

In [ ]:
df_reviews_parquet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59305 entries, 0 to 59304
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   User_Id             59305 non-null  object
 1   Item_Id             59305 non-null  int64 
 2   Recommend           59305 non-null  bool  
 3   Year                59305 non-null  int64 
 4   Date                59305 non-null  object
 5   Sentiment_Analysis  59305 non-null  int64 
dtypes: bool(1), int64(3), object(2)
memory usage: 2.3+ MB


## Data Set australian_users_items.json

In [ ]:
# Se inicializa una lista vacia donde alamcenare los datso del archivo .JSON
filas = []
# Se abre el archivo .JSON con el modulo 'with'
with open('E:\\AAADATOS\\Henry\\AA_Data_Science\\MATERIAL_PI\\DATA_PI MLOps - STEAM\\Data_sets\\australian_users_items.json', 'r', encoding='utf-8') as archivo:
    for line in archivo.readlines():
        filas.append(ast.literal_eval(line))

In [ ]:
# Se inicializa el DataFrame a partir de la lista donde se extrajo el archivo .JSON
df_items = pd.DataFrame(filas)
df_items

user_id  items_count           steam_id  \
0            76561197970982479          277  76561197970982479   
1                      js41637          888  76561198035864385   
2                    evcentric          137  76561198007712555   
3                   Riot-Punch          328  76561197963445855   
4                        doctr          541  76561198002099482   
...                        ...          ...                ...   
88305        76561198323066619           22  76561198323066619   
88306        76561198326700687          177  76561198326700687   
88307  XxLaughingJackClown77xX            0  76561198328759259   
88308        76561198329548331            7  76561198329548331   
88309        edward_tremethick            0  76561198331598578   

                                                user_url  \
0      http://steamcommunity.com/profiles/76561197970...   
1                   http://steamcommunity.com/id/js41637   
2                 http://steamcommunity.com/id/evcentric   
3                http://steamcommunity.com/id/Riot-Punch   
4                     http://steamcommunity.com/id/doctr   
...                                                  ...   
88305  http://steamcommunity.com/profiles/76561198323...   
88306  http://steamcommunity.com/profiles/76561198326...   
88307  http://steamcommunity.com/id/XxLaughingJackClo...   
88308  http://steamcommunity.com/profiles/76561198329...   
88309     http://steamcommunity.com/id/edward_tremethick   

                                                   items  
0      [{'item_id': '10', 'item_name': 'Counter-Strik...  
1      [{'item_id': '10', 'item_name': 'Counter-Strik...  
2      [{'item_id': '1200', 'item_name': 'Red Orchest...  
3      [{'item_id': '10', 'item_name': 'Counter-Strik...  
4      [{'item_id': '300', 'item_name': 'Day of Defea...  
...                                                  ...  
88305  [{'item_id': '413850', 'item_name': 'CS:GO Pla...  
88306  [{'item_id': '11020', 'item_name': 'TrackMania...  
88307                                                 []  
88308  [{'item_id': '304930', 'item_name': 'Unturned'...  
88309                                                 []  

[88310 rows x 5 columns]

Se evidencia que la columna "items" contiene listas como dato así que se tiene que proceder a expandirla

In [55]:
df_items = df_items.explode('items').reset_index() # Se redefine el DatFrame con una descomposición del campo para desanidarla y se reinicia el conteo del índice
df_items = df_items.drop(columns='index') # Se elimina el campo 'index'
df_items = pd.concat([df_items, pd.json_normalize(df_items['items'])], axis=1) # Se normalizan los datos JSON de la columna 'reviews' aplanando los datos, convirtiendolos en un formato tabular y concatenando de manera horizontal el DatFrame original con el del contendio de 'reviews' 
df_items.head(2)

user_id  items_count           steam_id  \
0  76561197970982479          277  76561197970982479   
1  76561197970982479          277  76561197970982479   
2  76561197970982479          277  76561197970982479   
3  76561197970982479          277  76561197970982479   
4  76561197970982479          277  76561197970982479   

                                            user_url  \
0  http://steamcommunity.com/profiles/76561197970...   
1  http://steamcommunity.com/profiles/76561197970...   
2  http://steamcommunity.com/profiles/76561197970...   
3  http://steamcommunity.com/profiles/76561197970...   
4  http://steamcommunity.com/profiles/76561197970...   

                                               items item_id  \
0  {'item_id': '10', 'item_name': 'Counter-Strike...      10   
1  {'item_id': '20', 'item_name': 'Team Fortress ...      20   
2  {'item_id': '30', 'item_name': 'Day of Defeat'...      30   
3  {'item_id': '40', 'item_name': 'Deathmatch Cla...      40   
4  {'item_id': '50', 'item_name': 'Half-Life: Opp...      50   

                   item_name  playtime_forever  playtime_2weeks  
0             Counter-Strike               6.0              0.0  
1      Team Fortress Classic               0.0              0.0  
2              Day of Defeat               7.0              0.0  
3         Deathmatch Classic               0.0              0.0  
4  Half-Life: Opposing Force               0.0              0.0

### Se empieza a analizar y transformar los datos del DataFrame

In [57]:
# Se verifica la información de los datos que contiene el DataFrame
df_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5170015 entries, 0 to 5170014
Data columns (total 9 columns):
 #   Column            Dtype  
---  ------            -----  
 0   user_id           object 
 1   items_count       int64  
 2   steam_id          object 
 3   user_url          object 
 4   items             object 
 5   item_id           object 
 6   item_name         object 
 7   playtime_forever  float64
 8   playtime_2weeks   float64
dtypes: float64(2), int64(1), object(6)
memory usage: 355.0+ MB


NULOS

In [58]:
# Conteo de Nulos
nulos = df_items.isna().sum()
nulos

user_id                 0
items_count             0
steam_id                0
user_url                0
items               16806
item_id             16806
item_name           16806
playtime_forever    16806
playtime_2weeks     16806
dtype: int64

In [60]:
# Se hace una operación matemática para determinar que poorcentaje de todo el DataSet representan esos valores Nulos
total_filas_df = 5170015
cant_nulos = 16806
print(f'La cantidad de valores nulos en el total del DataFrame es de {round(((cant_nulos/total_filas_df)*100), 2)}%')

La cantidad de valores nulos en el total del DataFrame es de 0.33%


Gracias a la operación anterior se determina que la cantidad de datos Nulos en el conjunto total del DataFrame representa una cantidad muy infima de los datos por lo cual se decide a ser eliminados

In [61]:
# Se eliminan los valores Nulos
df_items = df_items.dropna()

In [62]:
# Nuevamente se ha un conteo de Nulos
nulos = df_items.isna().sum()
nulos

user_id             0
items_count         0
steam_id            0
user_url            0
items               0
item_id             0
item_name           0
playtime_forever    0
playtime_2weeks     0
dtype: int64

In [63]:
# Se verifica nuevamente la información del DataFrame
df_items.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5153209 entries, 0 to 5170013
Data columns (total 9 columns):
 #   Column            Dtype  
---  ------            -----  
 0   user_id           object 
 1   items_count       int64  
 2   steam_id          object 
 3   user_url          object 
 4   items             object 
 5   item_id           object 
 6   item_name         object 
 7   playtime_forever  float64
 8   playtime_2weeks   float64
dtypes: float64(2), int64(1), object(6)
memory usage: 393.2+ MB


#### Eliminar campos no relevantes

In [65]:
# Se procede a eliminar campos no relevantes hasta este punto
df_items.drop(['user_url','items','playtime_2weeks'], axis=1, inplace=True)

C:\Users\DavidssonPC\AppData\Local\Temp\ipykernel_7448\2824205854.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_items.drop(["user_url","items"], axis=1, inplace=True)


Se considera el campo "playtime_2weeks" irrelevante ya que en las funciones que se especifican en la consigna solo piden las horas

#### Transformación de datos

Se procede a cambiar el tipo de dato del campo "item_id" por tipo int

In [68]:
# Se crea una variable que contiene los valores unicos del campo 'item_id' para evalular que solo contenga numeros
valores_item_id = df_items['item_id'].unique() 
valores_item_id

array(['10', '20', '30', ..., '354280', '433920', '485270'], dtype=object)

In [69]:
# Ahora que se sabe que solo contiene valores númericos se procede a realizar el cambio de tipo de dato
df_items['item_id'] = df_items['item_id'].astype(int)

C:\Users\DavidssonPC\AppData\Local\Temp\ipykernel_7448\1433158164.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_items['item_id'] = df_items['item_id'].astype(int)


In [70]:
df_items.info() # Se verifica que haya cambiado el tipo de dato del campo 'item_id'

<class 'pandas.core.frame.DataFrame'>
Index: 5153209 entries, 0 to 5170013
Data columns (total 7 columns):
 #   Column            Dtype  
---  ------            -----  
 0   user_id           object 
 1   items_count       int64  
 2   steam_id          object 
 3   item_id           int32  
 4   item_name         object 
 5   playtime_forever  float64
 6   playtime_2weeks   float64
dtypes: float64(2), int32(1), int64(1), object(3)
memory usage: 294.9+ MB


In [72]:
# Se procede a cambiar el tipo de dato del campo 'playtime_forever' y 'playtime_2weeks' a dato de tipo int ya que será necesario para un proceso futuro
df_items['playtime_forever'] = df_items['playtime_forever'].astype(int)
df_items['playtime_2weeks'] = df_items['playtime_2weeks'].astype(int)
df_items.info() # Se evalúa que efectivamente se hayan cambiado los tipos de datos

<class 'pandas.core.frame.DataFrame'>
Index: 5153209 entries, 0 to 5170013
Data columns (total 7 columns):
 #   Column            Dtype 
---  ------            ----- 
 0   user_id           object
 1   items_count       int64 
 2   steam_id          object
 3   item_id           int32 
 4   item_name         object
 5   playtime_forever  int32 
 6   playtime_2weeks   int32 
dtypes: int32(3), int64(1), object(3)
memory usage: 255.6+ MB


C:\Users\DavidssonPC\AppData\Local\Temp\ipykernel_7448\4126780482.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_items['playtime_forever'] = df_items['playtime_forever'].astype(int)
C:\Users\DavidssonPC\AppData\Local\Temp\ipykernel_7448\4126780482.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_items['playtime_2weeks'] = df_items['playtime_2weeks'].astype(int)


In [73]:
df_items

user_id  items_count           steam_id  item_id  \
0        76561197970982479          277  76561197970982479       10   
1        76561197970982479          277  76561197970982479       20   
2        76561197970982479          277  76561197970982479       30   
3        76561197970982479          277  76561197970982479       40   
4        76561197970982479          277  76561197970982479       50   
...                    ...          ...                ...      ...   
5170009  76561198329548331            7  76561198329548331   346330   
5170010  76561198329548331            7  76561198329548331   373330   
5170011  76561198329548331            7  76561198329548331   388490   
5170012  76561198329548331            7  76561198329548331   521570   
5170013  76561198329548331            7  76561198329548331   519140   

                             item_name  playtime_forever  playtime_2weeks  
0                       Counter-Strike                 6                0  
1                Team Fortress Classic                 0                0  
2                        Day of Defeat                 7                0  
3                   Deathmatch Classic                 0                0  
4            Half-Life: Opposing Force                 0                0  
...                                ...               ...              ...  
5170009                   BrainBread 2                 0                0  
5170010                    All Is Dust                 0                0  
5170011  One Way To Die: Steam Edition                 3                3  
5170012          You Have 10 Seconds 2                 4                4  
5170013                     Minds Eyes                 3                3  

[5153209 rows x 7 columns]

En este punto que ya se tienen los campos relevantes se procede a normalizar los nombre de estos

In [ ]:
df_items = df_items.rename(columns={'user_id': 'User_Id'})
df_items = df_items.rename(columns={'items_count': 'Items_Count'})
df_items = df_items.rename(columns={'steam_id': 'Steam_Id'})
df_items = df_items.rename(columns={'item_id': 'Item_Id'})
df_items = df_items.rename(columns={'item_name': 'Item_Name'})
df_items = df_items.rename(columns={'playtime_forever': 'Playtime_Forever_Hours'})
df_items.head(2)

User_Id Item_Id Recommend  \
0  76561197970982479    1250      True   
1  76561197970982479   22200      True   

                                              Review  Year       Date  
0  Simple yet with great replayability. In my opi...  2011 2011-11-05  
1               It's unique and worth a playthrough.  2011 2011-07-15

### Se guarda el DataFrame en formatos mas ligeros

In [ ]:
df_items.to_csv('user_items.csv', index=False) # Se guarda el DataFrame en formato CSV)

In [ ]:
# Se convierte el archivo CSV a parquet bajo la compresión "gzip"
ruta_csv_reviews = 'E:\\AAADATOS\\Henry\\AA_Data_Science\\MATERIAL_PI\\PI_ML_OPS_STEAM_DSFT17\\user_items.csv' # Se crea una variable con la ruta del archivo CSV
df_items_temp = pd.read_csv(ruta_csv_reviews) # Se lee ese CSV en un nuevo DataFrame temporal para segurar como estan los datos
df_items_temp

Se guarada el DataFrame temporal en formato comprimido gzip

In [ ]:
df_items_temp.to_csv('user_items.gzip', compression='gzip', index=False)

In [ ]:
# Se verifica que el archivo gzip comprimido pueda ser leido de manera efectiva y que no hayan problemas en los datos que este arroja
df_items_parquet = pd.read_csv('user_items.gzip', compression='gzip')
df_items_parquet

In [ ]:
df_items_parquet.info()